File: splitter/src/main/kotlin/com/github/telegram_bots/rss_manager/splitter/service/SubscriptionRepository.kt
package com.github.telegram_bots.rss_manager.splitter.service

import com.github.telegram_bots.rss_manager.splitter.domain.Subscription
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Repository
import java.sql.ResultSet

@Repository
class SubscriptionRepository(private val jdbc: JdbcTemplate) {
    fun getSubscriptions(channelLink: String, limit: Int, offset: Int): List<Subscription> = jdbc
        .query(
                """
                | SELECT s.name, u.telegram_id
                | FROM subscriptions AS s
                | JOIN channels AS c ON c.id = s.channel_id
                | JOIN users AS u ON u.id = s.user_id
                | WHERE c.url = ?
                | LIMIT ? OFFSET ?
                """.trimMargin(),
                arrayOf(channelLink, limit, offset),
                { rs: ResultSet, _ -> Subscription(rs) }
        )
}

File: core/src/main/scala/com/github/telegram_bots/telegram_rss_manager/core/persistence/SubscriptionRepository.scala
package com.github.telegram_bots.telegram_rss_manager.core.persistence

import com.github.telegram_bots.telegram_rss_manager.core.domain.Channel.ChannelID
import com.github.telegram_bots.telegram_rss_manager.core.domain.Subscription.SubscriptionName
import com.github.telegram_bots.telegram_rss_manager.core.domain.User.UserID
import com.github.telegram_bots.telegram_rss_manager.core.domain.{Channel, Subscription}
import com.github.telegram_bots.telegram_rss_manager.core.persistence.Mappers.channelMapper
import slick.jdbc.PostgresProfile.api._

import scala.concurrent.{ExecutionContext, Future}

class SubscriptionRepository(db: Database) {
  private implicit val executionContext: ExecutionContext = db.ioExecutionContext
  private val subscriptionQuery = TableQuery[Subscriptions]

  def subscribe(userId: UserID, channelId: ChannelID, name: SubscriptionName = "main"): Future[Int] = {
    val query = subscriptionQuery += Subscription(userId, channelId, name)

    db.run { query }
  }

  def unsubscribe(userId: UserID, channelId: ChannelID, name: SubscriptionName = "main"): Future[Int] = {
    val query = subscriptionQuery.filter(_.userId === userId)
      .filter(_.channelId === channelId)
      .filter(_.name === name)
      .delete

    db.run { query }
  }

  def list(userId: UserID): Future[Seq[Channel]] = {
    val query = sql"""
      SELECT c.*
      FROM subscriptions AS s
      JOIN channels AS c ON c.id = s.channel_id
      WHERE s.user_id = $userId
      """

    db.run { query.as[Channel] }
  }

  class Subscriptions(tag: Tag) extends Table[Subscription](tag, "subscriptions") {
    def userId: Rep[Int] = column[Int]("user_id")
    def channelId: Rep[Int] = column[Int]("channel_id")
    def name: Rep[String] = column[String]("name")

    def * = (userId, channelId, name) <> ((Subscription.apply _).tupled, Subscription.unapply)
  }
}